app/src/main/java/fr/neamar/kiss/dataprovider/AppProvider.java

```diff
@@ -19,30 +20,68 @@ public void reload() {
         ArrayList<Pojo> records = new ArrayList<>();
 
         int relevance;
-        int matchPositionStart;
-        int matchPositionEnd;
         String appNameNormalized;
 
-        final String queryWithSpace = " " + query;
         for (Pojo pojo : pojos) {
             relevance = 0;
             appNameNormalized = pojo.nameNormalized;
 
-            matchPositionEnd = 0;
-            if (appNameNormalized.startsWith(query)) {
-                relevance = 100;
-                matchPositionStart = 0;
-                matchPositionEnd = query.length();
-            } else if ((matchPositionStart = appNameNormalized.indexOf(queryWithSpace)) > -1) {
-                relevance = 50;
-                matchPositionEnd = matchPositionStart + queryWithSpace.length();
-            } else if ((matchPositionStart = appNameNormalized.indexOf(query)) > -1) {
-                relevance = 1;
-                matchPositionEnd = matchPositionStart + query.length();
+            ArrayList<Pair<Integer, Integer>> matchPositions = new ArrayList<>();
+
+            int queryPos = 0;
+            int appPos = 0;
+            int beginMatch = 0;
+            int matchedWordStarts = 0;
+            int totalWordStarts = 0;
+
+            boolean match = false;
+            for (char cApp : appNameNormalized.toCharArray()) {
+                if (queryPos < query.length() && query.charAt(queryPos) == cApp) {
+                    // If we aren't already matching something, let's save the beginning of the match
+                    if (!match) {
+                        beginMatch = appPos;
+                        match = true;
+                    }
+
+                    // If we are at the beginning of a word, add it to matchedWordStarts
+                    if (Character.isUpperCase(pojo.name.charAt(appPos)) || appPos == 0 || Character.isWhitespace(pojo.name.charAt(appPos - 1)))
```

Comment:
> OQ: do we want to enforce a capital letter for a first word?

There is no need to, `appPos == 0` is there to handle that.

> OQ: if appPos == 0, it may be worth boosting the app relevance too?

Well if `appPos == 0` is true, we already add 1 to `matchedWordStarts`. Imo this is not necessary because it will allow something like `Barcode Scanner` to have the same score if we write `Bar` or `Sca`, and I think it's a good thing.